[pytorch] cuBLAS addmm malfunction test

souravmandal opened the pull request:
Summary: Create unit test to detect cuBLAS breakage via large differences between CPU and GPU addmm invocations

Test Plan:
Sample unit test output --

[...]
test_cublas_addmm_size_10000_cpu_bfloat16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_10000_cpu_float16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_10000_cpu_float32 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_1000_cpu_bfloat16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_1000_cpu_float16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_1000_cpu_float32 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_100_cpu_bfloat16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_100_cpu_float16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_100_cpu_float32 (test_linalg.TestLinalgCPU) ... ok
[...]

Reviewed By: mikekgfb

Differential Revision: D39433029

lezcano reviewed (commented):
We are already testing quite extensively `matmul` in the tests `test_matmul_small_brute_force_{1,2,3}d_Nd` against NumPy. What's the reason for wanting to test `addmm` on its own?

souravmandal commented:
We (@mikekgfb  @ngimel) have observed in the past that cuBLAS breaks such that it produces random data, or crashes outright, esp. for large input arrays. It would be useful to test whether a new cuBLAS version triggers that.

Lezcano commented on test/test_linalg.py:
```diff
@@ -7453,6 +7454,54 @@ def test_preferred_linalg_library(self):
+        #
+        # Get dims
+        n, m, p = (size + 1, size, size + 2)
+        # Make random tensors on CPU
+        torch.manual_seed(123456789)
```
Why is this needed?

souravmandal replied:
For reproducibility of specific failed tests, and to enable comparison between dtypes.  However, we might also want to sample across different random matrices for each dtype/size combination.  In this case, it would be better to run each dtype/size combination several times, and having the same random test matrices matters less.

Lezcano replied:
I'm pretty sure this is already done at a test suit level.

Lezcano commented on test/test_linalg.py:
```diff
@@ -7453,6 +7454,54 @@ def test_preferred_linalg_library(self):
+        n, m, p = (size + 1, size, size + 2)
+        # Make random tensors on CPU
+        torch.manual_seed(123456789)
+        m_beta = make_tensor(1, dtype=dtype, device='cpu')
```
You might want to do
```python
make_arg = partial(make_tensor, dtype=dtype, device="cpu")
```

lezcano reviewed (commented):
In any case, I think it'd be better to user NumPy to compare against, as we are sure that it will return correct results.
Also, this way, you could simply skip `torch.float16` in CPU and the code could be heavily simplified. See for example how `check_single_matmul` does this, together with dynamic tolerances for robustness.

Lezcano commented:
Also, this issue may be relevant https://github.com/pytorch/pytorch/issues/84538. I think @srossross is looking into implementing this one.

souravmandal commented:
The issue with numpy is that it does not support bfloat16 ([ref1](https://data-apis.org/array-api/latest/purpose_and_scope.html?highlight=bfloat16#out-of-scope), [ref2](https://github.com/numpy/numpy/issues/19808)).  To simplify one could just make the reference tensors in each call float32, and apply the dtype to just the torch tensors on GPU.

mikekgfb commented:
Right, this is not numeric accuracy, this is ensuring that cuBLAS does not crash or produce wildly incorrect results.  As such, we do want to exercise the bfloat16, and compare with reasonable bounds to the epxected result obtained by computing with another numeric representation.

zrphercule reviewed (approved):
Stamp

ngimel commented on test/test_linalg.py:
```diff
@@ -7453,6 +7454,54 @@ def test_preferred_linalg_library(self):
+            m_2 = m_2.to(dtype=torch.float16)
+            res_cpu = res_cpu.to(dtype=torch.float16)
+        # Move tensors to GPU
+        m_beta = m_beta.to("cuda")
```
`.to(device)`, and similarly in all other places

souravmandal replied:
Since this test is specific to cuBLAS, presumably it should only run for a CUDA device; removing "device" param

ngimel commented on test/test_linalg.py:
```diff
@@ -7453,6 +7454,54 @@ def test_preferred_linalg_library(self):
+        m_2 = make_tensor((m, p), dtype=dtype, device='cpu')
+        # *FLOAT16 Special Handling*
+        # Backend does not tensorize float16 on CPU, so convert to float32
+        if dtype == torch.float16:
```
same approach can be used for bfloat16 (with larger error bars)

souravmandal committed:
[pytorch] cuBLAS addmm malfunction test (#85084)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85084

Create unit test to detect cuBLAS breakage via large differences between CPU and GPU addmm invocations

Test Plan:
Sample unit test output --

[...]
test_cublas_addmm_size_10000_cpu_bfloat16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_10000_cpu_float16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_10000_cpu_float32 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_1000_cpu_bfloat16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_1000_cpu_float16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_1000_cpu_float32 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_100_cpu_bfloat16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_100_cpu_float16 (test_linalg.TestLinalgCPU) ... ok
test_cublas_addmm_size_100_cpu_float32 (test_linalg.TestLinalgCPU) ... ok
[...]

Reviewed By: mikekgfb

Differential Revision: D39433029

fbshipit-source-id: b308ecceb44eab1afb039c98f4e1b6aa8ddb8f53

mikekgfb commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "broke tests on trunk, https://github.com/pytorch/pytorch/actions/runs/3098347639/jobs/5017166419" -m nosignal

clee2000 commented:
@pytorchbot revert -m "broke tests on trunk, https://github.com/pytorch/pytorch/actions/runs/3098347639/jobs/5017166419" -c nosignal

weiwangmeta commented:
This link instead: https://github.com/pytorch/pytorch/actions/runs/3098294186

malfet commented:
By the way, how much total test time this PR adds? (Though addmm even for 10kx10k matrices should be pretty quick )

malfet commented on test/test_linalg.py:
```diff
@@ -7453,6 +7454,59 @@ def test_preferred_linalg_library(self):
         self.assertEqual(out_ref, out1.cpu())
         self.assertEqual(out1, out2)
 
+    @onlyCUDA
```
You'll probably need `@unittest.skipIf(not CUDA11OrLater, "Only CUDA 11+ is supported")` unless you want to fix CUDA-10.2 bugs
